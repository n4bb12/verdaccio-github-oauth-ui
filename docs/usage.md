# Usage

1. [Login](#login)
   1. [On the Verdaccio UI](#on-the-verdaccio-ui)
   2. [On the Command Line](#on-the-command-line)
      1. [Option A) Using the Built-In CLI](#option-a-using-the-built-in-cli)
      2. [Option B) Using the Commands from the UI](#option-b-using-the-commands-from-the-ui)
2. [Signing Out](#signing-out)
   1. [On the Verdaccio UI](#on-the-verdaccio-ui-1)
   2. [On the Command Line](#on-the-command-line-1)

## Login

### On the Verdaccio UI

<img src="screenshots/authorize.png" width="398" />

- Click the login button. You get redirected to GitHub and prompted to authorize the registry to access your public information.
- If the registry requires private org membership, click the <kbd>Request</kbd> or <kbd>Grant</kbd> button next to the org to get `read:org` access to the registry.
- Authorize the registry to access your GitHub user and org info. You only need to do this once.
- Once completed, you'll be redirected back to Verdaccio.
- You are now logged in ðŸŽ‰.

### On the Command Line

#### Option A) Using the Built-In CLI

The easiest way to configure npm is to use this short command:

```bash
npx verdaccio-github-oauth-ui --registry http://localhost:4873
```

On success you should see this page in your browser:

<img src="screenshots/all-done.png" width="208" />

#### Option B) Using the Commands from the UI

- Verdaccio 5:

Open the "Configuration" dialog and click "Copy to clipboard":

<img src="screenshots/configuration-dialog.png" width="450" />

- Run the copied commands on your terminal:

```bash
npm config set //localhost:4873/:always-auth true
npm config set //localhost:4873/:_authToken "SECRET_TOKEN"
```

- Verify npm is set up correctly by running the `whoami` command:

```bash
npm whoami --registry http://localhost:4873
```

If you see your GitHub username, you are ready to start installing and
publishing packages.

## Signing Out

### On the Verdaccio UI

Click the <kbd>Logout</kbd> button as per usual.

### On the Command Line

Unless OAuth access is revoked in the GitHub settings, the token is valid
indefinitely.
