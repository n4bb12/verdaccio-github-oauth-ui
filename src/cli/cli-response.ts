import { logger } from "../logger"

export function respondWithCliMessage(status: string, message: string) {
  switch (status) {
    case "success":
      logger.log("All done! We've updated your npm configuration.")
      break

    default:
      logger.error(message)
      break
  }
}
