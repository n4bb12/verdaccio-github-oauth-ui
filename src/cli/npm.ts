import { execSync } from "child_process"
import minimist from "minimist"
import { URL } from "url"

const npmConfig = JSON.parse(execSync("npm config list --json").toString())

export function getRegistry() {
  const args = minimist(process.argv.slice(2))
  return (args.registry || npmConfig.registry).trim()
    .replace(/\/?$/, "") // Remove potential trailing slash
}

export function getConfigFile() {
  return npmConfig.userconfig
}

export function getSaveCommands(registry: string, token: string) {
  const url = new URL(registry)
  const baseUrl = url.host + url.pathname
  return [
    `npm config set //${baseUrl}:_authToken "${token}"`, // lgtm [js/command-line-injection]
    `npm config set //${baseUrl}:always-auth true`,
  ]
}

export function save(registry: string, token: string) {
  getSaveCommands(registry, token).forEach(command => execSync(command))
}
