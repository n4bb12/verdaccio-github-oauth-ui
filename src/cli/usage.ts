import { pluginName } from "../constants"
import { logger } from "../logger"

export function getUsageInfo() {
  return [
    "It seems you are using the default npm registry.",
    "Please update it to your Verdaccio URL by either running:",
    "",
    "npm config set registry <URL>",
    "",
    "or by using the registry argument",
    "",
    `npx verdaccio-${pluginName} --registry <URL>`,
  ]
}

export function printUsage() {
  getUsageInfo().forEach(line => logger.log(line))
}
