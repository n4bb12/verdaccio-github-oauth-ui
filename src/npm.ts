import { execSync } from "child_process"
import minimist from "minimist"
import { URL } from "url"

import { logger } from "./logger"

let npmConfig: any

function parseCliArgs() {
  return minimist(process.argv.slice(2))
}

function runCommand(command: string) {
  logger.log(`Running command: ${command}`)
  return execSync(command)
}

function getNpmConfig() {
  if (!npmConfig) {
    npmConfig = JSON.parse(runCommand("npm config list --json").toString())
  }
  return npmConfig
}

function removeTrailingSlash(input: string) {
  return input.trim().replace(/\/?$/, "")
}

export function getRegistryUrl() {
  const cliArgs = parseCliArgs()

  const registry = cliArgs.registry || getNpmConfig().registry

  return removeTrailingSlash(registry)
}

export function getNpmConfigFile() {
  return getNpmConfig().userconfig
}

export function getNpmSaveCommands(registry: string, token: string) {
  const url = new URL(registry)
  const baseUrl = url.host + url.pathname

  return [
    `npm config set //${baseUrl}:always-auth true`,
    `npm config set //${baseUrl}:_authToken "${token}"`,
  ]
}

export function saveNpmToken(token: string) {
  const registry = getRegistryUrl()
  const commands = getNpmSaveCommands(registry, token)

  commands.forEach((command) => runCommand(command))
}
