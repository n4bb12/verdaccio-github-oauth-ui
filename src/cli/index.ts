import express from "express"
import open from "open"

import { buildStatusPage } from "../statusPage"
import { getConfigFile, getRegistry, save } from "./npm"
import { printUsage } from "./usage"

const registry = getRegistry()

if (registry.includes("registry.npmjs.org")) { // lgtm [js/incomplete-url-substring-sanitization]
  printUsage()
  process.exit(1)
}

open(registry + "/oauth/authorize")

const successPage = buildStatusPage(`
  <h1>All done!</h1>
  <p>We've updated your npm configuration.</p>
  <p><code>${getConfigFile()}</code></p>
`)

const server = express()
  .get("/", (req, res) => {
    const token = decodeURIComponent(req.query.token)
    save(registry, token)
    res.setHeader("Content-Type", "text/html")
    res.send(successPage)
    server.close()
  })
  .listen(8239)
