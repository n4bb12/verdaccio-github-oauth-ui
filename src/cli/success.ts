import { pluginName } from "../server/plugin/Config"
import { npmConfig } from "./npm"

export const successPage = `<!DOCTYPE html>
<html>
  <head>
    <title>${pluginName}</title>
    <style>
      html,
      body {
        padding: 0;
        margin: 0;
        height: 100%;
        background-color: #e0e0e0;
        color: #24292F;
        font-family: Helvetica, sans-serif;
        position: relative;
        text-align: center;
      }
      .wrap {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
      }
      img {
        filter: drop-shadow(0 0.5rem 0.5rem #24292F80);
      }
    </style>
  </head>
  <body>
    <div class="wrap">
      <img src="https://verdaccio.org/docs/assets/logo/symbol/svg/verdaccio-blackwhite.svg" alt="verdaccio-blackwhite.svg" />
      <h1>All done!</h1>
      <p>We've updated your npm configuration.</p>
      <p><code>${npmConfig.userconfig}</code></p>
    </div>
  </body>
</html>
`
