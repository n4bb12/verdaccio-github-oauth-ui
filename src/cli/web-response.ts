import { Response } from "express"
import { getNpmConfigFile } from "../npm"

import { buildErrorPage, buildStatusPage } from "../statusPage"

const withBackLink = false

const successPage = buildStatusPage(
  `<h1>All done!</h1>
  <p>We've updated your npm configuration.</p>
  <p><code>${getNpmConfigFile()}</code></p>`,
  withBackLink,
)

export function respondWithWebPage(
  status: string,
  message: string,
  res: Response,
) {
  res.setHeader("Content-Type", "text/html")

  switch (status) {
    case "success":
      res.status(200)
      res.send(successPage)
      break

    default:
      res.status(500)
      res.send(buildErrorPage(message, withBackLink))
      break
  }
}
