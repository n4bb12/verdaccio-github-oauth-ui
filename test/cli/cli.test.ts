import { getConfigFile, getRegistry, getSaveCommands } from "src/cli/npm"
import { getUsageInfo, printUsage } from "src/cli/usage"
import { testBaseUrl, testNPMToken } from "test/utils"

describe("CLI", () => {

  const argv = process.argv

  describe("usage", () => {
    it("usage info matches the snapshot", () => {
      expect(getUsageInfo()).toMatchSnapshot()
    })

    it("does not throw an error", () => {
      printUsage()
    })
  })

  describe("npm", () => {
    it("successfully finds the npm configuration", () => {
      expect(getRegistry()).toBeTruthy()
      expect(getConfigFile()).toBeTruthy()
    })

    it("uses the correct registry", () => {
      process.argv = argv
      const first = getRegistry()
      process.argv = [...argv, "--registry", testBaseUrl]
      const second = getRegistry()

      expect(first).toBeTruthy()
      expect(second).toBe(testBaseUrl)
      expect(first).not.toBe(second)
    })

    it("removes trailing slashes", () => {
      process.argv = [...argv, "--registry", "https://my.registry.com/"]
      expect(getRegistry()).toBe("https://my.registry.com")
    })

    it("save commands match the snapshot", () => {
      expect(getSaveCommands(testBaseUrl, testNPMToken)).toMatchSnapshot()
    })
  })

})
