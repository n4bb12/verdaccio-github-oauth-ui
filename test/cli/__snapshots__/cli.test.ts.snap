// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI npm save commands match the snapshot 1`] = `
Array [
  "npm config set //localhost:4873/:_authToken \\"test-npm-token\\"",
  "npm config set //localhost:4873/:always-auth true",
]
`;

exports[`CLI usage usage info matches the snapshot 1`] = `
Array [
  "It seems you are using the default npm registry.",
  "Please update it to your Verdaccio URL by either running:",
  "",
  "npm config set registry <URL>",
  "",
  "or by using the registry argument",
  "",
  "npx verdaccio-github-oauth-ui --registry <URL>",
]
`;
